src/dilithium.cpp

```diff
@@ -0,0 +1,448 @@
+        throw MoCOCrWException("Cannot read nested public key ASN.1 structure");
+    }
+
+    KeyTypes keyType = getKeyTypeFromDilithiumForm(nestedPubKey->dilithiumForm);
+
+    return std::make_shared<DilithiumKeyImpl>(
+            std::vector<uint8_t>(nestedPubKey->pubKey->data,
+                                 nestedPubKey->pubKey->data + nestedPubKey->pubKey->length),
+            keyType);
+}
+
+std::shared_ptr<DilithiumKeyImpl> DilithiumKeyImpl::parseAsn1PrivateKey(
+        const std::vector<uint8_t> &pkcs8PrivKey)
+{
+    // Parse the RFC 5958 ASN.1 DER form
+    auto p8inf = openssl::_SSL_d2i_PKCS8_PRIV_KEY_INFO(pkcs8PrivKey.data(), pkcs8PrivKey.size());
+
+    const unsigned char *pk;
+    int ppklen;
+    // get the nested octet string
+    // we can use get0 here, as pk is not used outside of this function
+    // (d2i_dilithium_priv and the underlying ASN1_item_d2i_ex do an alloc according to the manpage)
+    if (!PKCS8_pkey_get0(nullptr, &pk, &ppklen, nullptr, p8inf.get())) {
+        throw MoCOCrWException("Cannot get private key data.\n");
+    }
+
+    // parse the nested octet string
+    DILITHIUM_PRIV_Ptr nestedPrivKey(d2i_dilithium_priv(nullptr, &pk, ppklen));
+    // As we cannot use OpensslCallPtr as this is defined in openssl_wrap.cpp we need to manually
+    // check the return value
+    if (!nestedPrivKey) {
+        throw MoCOCrWException("Cannot read nested private key ASN.1 structure");
+    }
+
+    KeyTypes keyType = getKeyTypeFromDilithiumForm(nestedPrivKey->dilithiumForm);
+
+    return std::make_shared<DilithiumKeyImpl>(
+            std::vector<uint8_t>(nestedPrivKey->privKey->data,
+                                 nestedPrivKey->privKey->data + nestedPrivKey->privKey->length),
+            keyType);
+}
+
+DilithiumKeyImpl DilithiumKeyImpl::getPublicKey() const
+{
+    if (!isPrivateKey()) {
+        return *this;
+    }
+    std::vector<uint8_t> public_key(getPubKeySize(_keyType));
+    std::function<int(uint8_t *, const uint8_t *)> publicKeyGetterFunc;
+    switch (_keyType) {
+        case KeyTypes::DILITHIUM2:
+            publicKeyGetterFunc = pqcrystals_dilithium2_ref_keypair_public_from_private;
+            break;
+        case KeyTypes::DILITHIUM3:
+            publicKeyGetterFunc = pqcrystals_dilithium3_ref_keypair_public_from_private;
+            break;
+        case KeyTypes::DILITHIUM5:
+            publicKeyGetterFunc = pqcrystals_dilithium5_ref_keypair_public_from_private;
+            break;
+        default:
+            throw MoCOCrWException("Unsupported key type");
+    }
+    if (publicKeyGetterFunc(public_key.data(), _key_data.data())) {
+        throw MoCOCrWException("Dilithium: Cannot get public key from private key.");
+    }
+    return DilithiumKeyImpl(public_key, _keyType);
+}
+
+bool DilithiumKeyImpl::hasValidKeySize() const
+{
+    if (getKeySize() != getPrivKeySize(_keyType) && getKeySize() != getPubKeySize(_keyType)) {
+        return false;
+    }
+    return true;
```

Comment:
```suggestion
    return getkeySize() == getPrivKeySize(_keyType) || getKeySize() == getPubkeySize(_keyType)
```
However, I think I already commented this in the header. I find a function that does a limited syntactical check like that one here rather weird. One should at least know if one wants a public or a private key. This information should be available in all scopes when checking for the correct size is relevant?